File: OOP_Encapsulation_Valentin_Arnautski/OOP_Encapsulation/Student.java
package OOP_Encapsulation;

public class Student {
    private String name;
    private int age;
    private String[] courses;

    public Student(String name, int age) {
        this.name = name;
        this.age = age;
        this.courses = new String[0];
    }

    public void addCourse(String course) {
        String[] newCourses = new String[courses.length + 1];
        for (int i = 0; i < courses.length; i++) {
            newCourses[i] = courses[i];
        }
        newCourses[courses.length] = course;
        courses = newCourses;
    }

    public void removeCourse(String course) {
        int index = -1;
        for (int i = 0; i < courses.length; i++) {
            if (courses[i].equals(course)) {
                index = i;
                break;
            }
        }
        if (index != -1) {
            String[] newCourses = new String[courses.length - 1];
            for (int i = 0, j = 0; i < courses.length; i++) {
                if (i != index) {
                    newCourses[j++] = courses[i];
                }
            }
            courses = newCourses;
        }
    }

    public String[] getCourses() {
        return courses;
    }

    public int getAge() {
        return age;
    }

    public String getName() {
        return name;
    }
}

File: OOP_Encapsulation_MilchoKasmetov/Student.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using static System.Net.Mime.MediaTypeNames;

namespace OOP_Encapsulation_MilchoKasmetov
{
    public class Student
    {
        private const int MinAge = 0;
        private const int MaxAge = 120;

        private string _name;
        private int _age;

        public Student(string name, int age)
        {
            this.Name = name;
            this.Age = age;
            this.Courses = new List<string>();
        }


        private string Name
        {
            get
            {
                return _name;
            }
            set
            {
                if (string.IsNullOrWhiteSpace(value) || String.IsNullOrEmpty(value))
                {
                    throw new Exception("Name cannot be empty.");
                }


                this._name = value;
            }
        }

        private int Age
        {
            get
            {
                return _age;
            }
            set
            {
                if (value < MinAge || value > MaxAge)
                {
                    throw new Exception($"Age should be between {MinAge} and {MaxAge}.");
                }


                this._age = value;
            }
        }

        public ICollection<string> Courses { get; set; }


        //addCourse(course: String) : adds a course to the courses array.
        //removeCourse(course: String): removes a course from the courses array.
        //getCourses() -> String[]: returns the courses array.
        //getAge() -> int: returns the age attribute.
        //getName() -> String: returns the name attribute.

        public void AddCourse(string course)
        {
            if (!this.Courses.Contains(course))
            {
                Courses.Add(course);

                Console.WriteLine("Succesfuly added course");
            }
            else
            {
                Console.WriteLine("Already in this course !");
            }
        }

        public void RemoveCourse(string course)
        {
            if (this.Courses.Contains(course))
            {
                Courses.Remove(course);

                Console.WriteLine("Succesfuly removed course");
            }
            else
            {
                Console.WriteLine("There is no such course !");
            }
        }

        public ICollection<string> GetCourses() => this.Courses;
        public int GetAge() => this.Age;
        public string GetName() => this.Name;
    }
}
